### ISSUE TITLE
[rllib] Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False

### ISSUE THREAD

**Issue by janblumenkamp** [ray-project/ray#9181]
### What is the problem?

Training on a GPU and restoring on non-GPU devices fails. This problem occurs when training with a GPU but attempting to restore on a CPU only device with torch. It can be reproduced as shown below, but I also noticed it with setups where evaluation is enabled.

```
Traceback (most recent call last):
  File "restore_test.py", line 21, in <module>
    trainer.restore(checkpoint_file)
  File "[...]/ray/tune/trainable.py", line 417, in restore
    self._restore(checkpoint_path)
  File "[...]/ray/rllib/agents/trainer.py", line 665, in _restore
    self.__setstate__(extra_data)
  File "[...]/ray/rllib/agents/trainer_template.py", line 187, in __setstate__
    Trainer.__setstate__(self, state)
  File "[...]/ray/rllib/agents/trainer.py", line 1095, in __setstate__
    self.workers.local_worker().restore(state["worker"])
  File "[...]/ray/rllib/evaluation/rollout_worker.py", line 850, in restore
    objs = pickle.loads(objs)
  File "[...]/torch/storage.py", line 136, in _load_from_bytes
    return torch.load(io.BytesIO(b))
  File "[...]/torch/serialization.py", line 593, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
  File "[...]/torch/serialization.py", line 773, in _legacy_load
    result = unpickler.load()
  File "[...]/torch/serialization.py", line 729, in persistent_load
    deserialized_objects[root_key] = restore_location(obj, location)
  File "[...]/torch/serialization.py", line 178, in default_restore_location
    result = fn(storage, location)
  File "[...]/torch/serialization.py", line 154, in _cuda_deserialize
    device = validate_cuda_device(location)
  File "[...]/torch/serialization.py", line 138, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.

```

*Ray version and other system information (Python version, TensorFlow version, OS):*

- Ray 0.8.6
- Python 3.7.6
- torch 1.5.0
- Ubuntu 18.04

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

Training (run on device with GPU):
```
from ray import tune
from ray.rllib.agents.ppo import PPOTrainer
tune.run(
    PPOTrainer,
    checkpoint_freq=1,
    config={
        "framework": "torch",
        "num_workers": 8,
        "num_gpus": 1,
        "env": "CartPole-v0"
    },
    stop={
        "training_iteration":1
    }
)
```

Restore (copy checkpoint to different device without GPU, adapt checkpoint path, run):
```
from ray.rllib.agents.ppo import PPOTrainer
import os
import json
import ray

ray.init()

checkpoint_path = "./ray_results/PPO/PPO_CartPole-v0_0_2020-06-28_09-58-597hejbdm7/checkpoint_10"

with open(checkpoint_path + '/../params.json') as json_file:
    checkpoint_config = json.load(json_file)

checkpoint_config["num_workers"] = 0
checkpoint_config["num_gpus"] = 0

trainer = PPOTrainer(
    env=checkpoint_config['env'],
    config=checkpoint_config
)
checkpoint_file = checkpoint_path + '/checkpoint-' + os.path.basename(checkpoint_path).split('_')[-1]
trainer.restore(checkpoint_file)
```

If we cannot run your script, we cannot fix your issue.

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html). (I get a completely different error if I try to reproduce this on the current master...)

**Comment by janblumenkamp** [comment]
A hack to locally resolve this is changing the function `_load_from_bytes` in `torch/storage.py` as follows:
```
def _load_from_bytes(b):
    return torch.load(io.BytesIO(b), map_location=torch.device('cpu'))
```

**Comment by janblumenkamp** [comment]
This is related to https://github.com/pytorch/pytorch/issues/16797

**Comment by rkooo567** [comment]
cc @sven1977

**Comment by lasagnaphil** [comment]
A bugfix for this issue was merged recently (#11208), can you check this again with the latest builds?
